src/lib/ndpi_main.c

```diff
@@ -1954,7 +1954,8 @@ struct ndpi_detection_module_struct *ndpi_init_detection_module(ndpi_init_prefs
 
   if(ndpi_str == NULL) {
 #ifdef NDPI_ENABLE_DEBUG_MESSAGES
-    NDPI_LOG_ERR(ndpi_str, "ndpi_init_detection_module initial malloc failed for ndpi_str\n");
+    /* We can't use NDPI_LOG* functions yet! */
+    fprintf(stderr, "ndpi_init_detection_module initial malloc failed for ndpi_str\n");
```

Comment:
A shared object should never assume that file descriptor 2 is a stderr fd, it could be anything, a fd to /dev/null, a socket, pipe or whatever fd.

nDPI already has a quite nice function called `ndpi_debug_printf`, which callback can be altered with `set_ndpi_debug_function`

Reply:
AFAIU, it is a little bit more complicated.
This log should be reported at the beginning of ndpi_init_detection_module(), which initializes the library and must be the first function called by the user.
At that point:
* you don't have a detection_module context, so you can't log anything via NDPI_LOG* macros
* the user can't have already called set_ndpi_debug_function(), so a custom log function isn't available yet. We could use the default implementation of ndpi_debug_printf(), but it prints on stdout. Your valid concern about stderr, apply to stdout, too.

Since this allocation error should be uncommon (to say the least), I think we have only two options:
1) don't log anything
2) directly print on stdout or stderr, having such printf protected via NDPI_ENABLE_DEBUG_MESSAGES define

I have no strong opinion one way or the other

Reply:
I see the issue. We could add an "early-logging-callback" as parameter for `ndpi_init_detection_module`.
However this would be an api break.